File: 0216_med_med_combination-sum-iii/solution.py
from typing import List


class Solution:

    def __init__(self) -> None:
        self.result = []
        self.path = []
        self.sum = 0

    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        self.back_tracking(k, n, 1)
        return self.result

    def back_tracking(self, k: int, n: int, start_index: int) -> None:
        if(len(self.path) == k):
            if self.sum == n:
                self.result.append(self.path[:])
            return
        
        for i in range(start_index, 9 - (k - len(self.path)) + 2):
            self.path.append(i)
            self.sum += i
            self.back_tracking(k, n, i + 1)
            self.path.pop()
            self.sum -= i

File: 0216_med_med_combination-sum-iii/solution.java
import java.util.*;


class Solution {
    List<List<Integer>> result = new ArrayList<>();
    LinkedList<Integer> path = new LinkedList<>();
    int sum = 0;

    public List<List<Integer>> combinationSum3(int k, int n) {
        backTracking(k, n, 1);
        return result;
    }

    private void backTracking(int k, int n, int startIndex) {
        if (path.size() == k) {
            if (sum == n) {
                result.add(new ArrayList<>(path));
            }
            return;
        }

        for (int i = startIndex; i <= 9 - (k - path.size()) + 1; i++) {
            path.add(i);
            sum += i;
            backTracking(k, n, i + 1);
            path.removeLast();
            sum -= i;
        }
    }
}